pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,778 @@
+			PALLET_STORAGE_PREFIX,
+			ITEMIZED_STORAGE_PREFIX,
+			&key,
+		)
+		.map_err(|_| Error::<T>::CorruptedState)?
+		.unwrap_or_default();
+
+		let prev_content_hash = existing_page.get_hash();
+		ensure!(target_hash == prev_content_hash, Error::<T>::StalePageState);
+
+		let updated_page =
+			ItemizedOperations::<T>::apply_item_actions(&existing_page, &actions[..]).map_err(
+				|e| match e {
+					PageError::ErrorParsing(err) => {
+						log::warn!(
+							"failed parsing Itemized msa={:?} schema_id={:?} {:?}",
+							state_owner_msa_id,
+							schema_id,
+							err
+						);
+						Error::<T>::CorruptedState
+					},
+					_ => Error::<T>::InvalidItemAction,
+				},
+			)?;
+
+		match updated_page.is_empty() {
+			true => {
+				StatefulChildTree::<T::KeyHasher>::kill(
+					&state_owner_msa_id,
+					PALLET_STORAGE_PREFIX,
+					ITEMIZED_STORAGE_PREFIX,
+					&key,
+				);
+				Self::deposit_event(Event::ItemizedPageDeleted {
+					msa_id: state_owner_msa_id,
+					schema_id,
+					prev_content_hash,
+				});
+			},
+			false => {
+				StatefulChildTree::<T::KeyHasher>::write(
+					&state_owner_msa_id,
+					PALLET_STORAGE_PREFIX,
+					ITEMIZED_STORAGE_PREFIX,
+					&key,
+					&updated_page,
+				);
+				Self::deposit_event(Event::ItemizedPageUpdated {
+					msa_id: state_owner_msa_id,
+					schema_id,
+					curr_content_hash: updated_page.get_hash(),
+					prev_content_hash,
+				});
+			},
+		};
+		Ok(())
+	}
+
+	/// Modifies a page from paginated storage by provided new page
+	///
+	/// # Events
+	/// * [`Event::PaginatedPageUpdated`]
+	///
+	fn modify_paginated(
+		state_owner_msa_id: MessageSourceId,
+		schema_id: SchemaId,
+		page_id: PageId,
+		target_hash: PageHash,
+		new_page: PaginatedPage<T>,
+	) -> DispatchResult {
+		let keys: PaginatedKey = (schema_id, page_id);
+		let existing_page: PaginatedPage<T> = StatefulChildTree::<T::KeyHasher>::try_read(
+			&state_owner_msa_id,
+			PALLET_STORAGE_PREFIX,
+			PAGINATED_STORAGE_PREFIX,
+			&keys,
+		)
+		.map_err(|_| Error::<T>::CorruptedState)?
+		.unwrap_or_default();
+
+		let prev_content_hash: PageHash = existing_page.get_hash();
+		ensure!(target_hash == prev_content_hash, Error::<T>::StalePageState);
+
+		StatefulChildTree::<T::KeyHasher>::write(
+			&state_owner_msa_id,
+			PALLET_STORAGE_PREFIX,
+			PAGINATED_STORAGE_PREFIX,
+			&keys,
+			&new_page,
+		);
+		Self::deposit_event(Event::PaginatedPageUpdated {
+			msa_id: state_owner_msa_id,
+			schema_id,
+			page_id,
+			curr_content_hash: new_page.get_hash(),
+			prev_content_hash,
+		});
+		Ok(())
+	}
+
+	/// Deletes a page from paginated storage
+	///
+	/// # Events
+	/// * [`Event::PaginatedPageDeleted`]
+	///
+	fn delete_paginated(
+		state_owner_msa_id: MessageSourceId,
+		schema_id: SchemaId,
+		page_id: PageId,
+		target_hash: PageHash,
+	) -> DispatchResult {
+		let keys: PaginatedKey = (schema_id, page_id);
+		if let Some(existing_page) =
+			StatefulChildTree::<T::KeyHasher>::try_read::<_, PaginatedPage<T>>(
```

Comment:
We use this in a couple places. I think it would be a good idea to make helper methods.